src/opensynth/models/faraday/gaussian_mixture/gmm_lightning.py

```diff
@@ -120,6 +125,9 @@ def training_step(self, batch: torch.Tensor) -> None:
             )
         responsibilities = log_responsibilities.exp()
 
+        # ensure the lowest cluster probability is 1/N.
+        responsibilities += 1 / (self.num_datapoints)
```

Comment:
I think this should be batch size rather than total size of dataset though because:

1. We also want to apply GMM to 'streaming dataset' where data cannot be load into memory. In streaming paradigm, the number of dataset will be known.

2. For a very very big dataset, we 1/N will be very low threshold (causing potential overfitting issue). Would be better I think to make this 1 / (BatchSize) - then end user just need to specify batch size as a thing to experiment/ tune (which is a standard ML tuning parameter).

Reply:
We should either:

1. Make this 1 / batch size and enforce num_clusters <= batch_size (like sklearn) _or_

2. Relax num_clusters <= batch_size and allow for arbitrary batch size, but we need learning rate especially for cases where batch_size <= num_clusters and batches with missing data could lead to previously learned weights being completely overridden.

Reply:
I went with 1, I didn't need to add an explicit check for `num_clusters <= batch_size` because when training the GMM we also run kmeans with sklearn, which checks that num clusters < batch size
The following error is raised when I tried running with n_clusters > n_samples
`ValueError: n_samples=5000 should be >= n_clusters=5001.`